sender_pool.go

```diff
@@ -146,35 +171,49 @@ func (p *LineSenderPool) Acquire(ctx context.Context) (LineSender, error) {
 				return nil, errors.New("tcp/s not supported for pooled senders, use http/s only")
 			}
 		}
-		return newHttpLineSender(conf)
+		s, err = newHttpLineSender(conf)
+	}
+
+	if err != nil {
+		return nil, err
 	}
 
+	p.numSenders++
+
+	ps := &pooledSender{
+		pool:    p,
+		wrapped: s,
+	}
+	atomic.AddUint64(&ps.tick, 1)
+	return ps, nil
 }
 
-// Release flushes the LineSender and returns it back to the pool. If the pool
-// is full, the sender is closed and discarded. In cases where the sender's
-// flush fails, it is not added back to the pool.
-func (p *LineSenderPool) Release(ctx context.Context, s LineSender) error {
-	// If there is an error on flush, do not add the sender back to the pool
-	if err := s.Flush(ctx); err != nil {
-		return err
+func (p *LineSenderPool) free(ctx context.Context, ps *pooledSender) error {
+	var flushErr error
+
+	if ps.dirty {
+		flushErr = ps.Flush(ctx)
 	}
 
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
-	for i := range p.senders {
-		if p.senders[i] == s {
-			return fmt.Errorf("LineSender %p has already been released back to the pool", s)
-		}
+	if flushErr != nil {
+		// Failed to flush, close and call it a day
+		p.numSenders--
+		closeErr := ps.wrapped.Close(ctx)
+		return fmt.Errorf("%s %w", flushErr, closeErr)
```

Comment:
I think you should check to see if `closeErr != nil` before wrapping it, otherwise it looks a bit messy when you print the error:

https://go.dev/play/p/EaIgaymmIHe

Reply:
Good spot. Done in 1e5624588cd3f29cb347fdc462eb8e1ca4ae4377